caluma/form/validators.py

```diff
@@ -286,34 +282,36 @@ def _get_answer_value(self, answer, document, parent):
     def validate_required(self, document, answer_tree):
         required_but_empty = []
         required_state = {}
-        for question in document.form.questions.all():
+        for question in document.form.all_questions().values(
+            "slug", "is_required", "is_hidden"
+        ):
             # TODO: can we iterate over questions of answers via answer_tree?
             try:
                 expr = "is_required"
                 is_required = (
-                    jexl.QuestionJexl(answer_tree, document.root_form.slug).evaluate(
-                        question.is_required
+                    jexl.QuestionJexl(answer_tree, document.form.slug).evaluate(
+                        question["is_required"]
                     )
                     and self.do_check_required
                 )
 
                 expr = "is_hidden"
-                is_hidden = jexl.QuestionJexl(
-                    answer_tree, document.root_form.slug
-                ).evaluate(question.is_hidden)
+                is_hidden = jexl.QuestionJexl(answer_tree, document.form.slug).evaluate(
+                    question["is_hidden"]
+                )
                 if is_required and not is_hidden:
-                    if answer_tree.get(question.slug, None) in EMPTY_VALUES:
-                        required_but_empty.append(question.slug)
+                    if answer_tree.get(question["slug"], None) in EMPTY_VALUES:
```

Comment:
I'd work on the naming a bit: In some places, `answer_tree` is a list, in others, it's a dict.